runtime/vm/BytecodeInterpreter.hpp

```diff
@@ -7301,10 +7306,77 @@ done:;
 	VMINLINE bool
 	isSubstitutable(j9object_t lhs, j9object_t rhs)
 	{
-		/*
-		 * TODO: this will be updated in a future PR.
-		 */
-		return false;
+		UDATA const objectHeaderSize = J9VMTHREAD_OBJECT_HEADER_SIZE(_currentThread);
+		J9Class * lhsClass = J9OBJECT_CLAZZ(_currentThread, lhs);
+		U_32 walkFlags = J9VM_FIELD_OFFSET_WALK_INCLUDE_STATIC | J9VM_FIELD_OFFSET_WALK_INCLUDE_INSTANCE;
+		J9ROMFieldOffsetWalkState state;
+		J9ROMFieldOffsetWalkResult *result = fieldOffsetsStartDo(_vm, lhsClass->romClass, VM_VMHelpers::getSuperclass(lhsClass), &state, walkFlags, lhsClass->flattenedClassCache);
+
+		while (NULL != result->field) {
+			J9UTF8 *signature = J9ROMNAMEANDSIGNATURE_SIGNATURE(&result->field->nameAndSignature);
+			U_8 sigChar = J9UTF8_DATA(signature)[0];
+
+			switch (sigChar) {
+			case 'Z': // Boolean
+			case 'B': // Byte
+			case 'C': // Char
+			case 'I': // Int
+			case 'S': { // Short 
+				I_32 lhsValue = *(I_32 *)((UDATA)lhs + objectHeaderSize + result->offset);
+				I_32 rhsValue = *(I_32 *)((UDATA)rhs + objectHeaderSize + result->offset);
+				if (lhsValue != rhsValue) {
+					return false;
+				}
+				break;
+			}
+			case 'J': { // Long
+				I_64 lhsValue = *(I_64 *)((UDATA)lhs + objectHeaderSize + result->offset);
+				I_64 rhsValue = *(I_64 *)((UDATA)rhs + objectHeaderSize + result->offset);
+				if (lhsValue != rhsValue) {
+					return false;
+				}
+				break;
+			}
+			case 'D': { // Double
+				jdouble *lhsValue = (jdouble *)((UDATA)lhs + objectHeaderSize + result->offset);
+				jdouble *rhsValue = (jdouble *)((UDATA)rhs + objectHeaderSize + result->offset);
+				I_32 result = helperDoubleCompareDouble(lhsValue, rhsValue);
+				if (result != 0) {
+					return false;
+				}
+				break;
+			}
+			case 'F': { // Float
+				jfloat *lhsValue = (jfloat *)((UDATA)lhs + objectHeaderSize + result->offset);
+				jfloat *rhsValue = (jfloat *)((UDATA)rhs + objectHeaderSize + result->offset);
+				I_32 result = helperFloatCompareFloat(lhsValue, rhsValue);
+				if (result != 0) {
+					return false;
+				}
+				break;
+			}
+			case '[': // Array
+			case 'L': // Nullable class type or interface type
+			case 'Q': { // Null-free class type
```

Comment:
for Q types you first need to determine if it has the `FastSubstitutableFlag` set

Reply:
you also need to determine if its flattened